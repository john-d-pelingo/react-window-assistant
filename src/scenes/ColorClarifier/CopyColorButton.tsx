import { ClickAwayListener, IconButton, Tooltip } from '@material-ui/core'
import { FileCopyRounded } from '@material-ui/icons'
import React, { FC, useState } from 'react'

import { copyToClipboard } from 'helpers/copyToClipboard'

interface ICopyColorButtonProps {
  color: { colorString: string; name: string }
}

export const CopyColorButton: FC<ICopyColorButtonProps> = ({ color }) => {
  const [isToolTipOpen, setIsToolTipOpen] = useState(false)

  const handleTooltipClose = () => {
    setIsToolTipOpen(false)
  }

  const handleButtonClick = () => {
    setIsToolTipOpen(true)
    copyToClipboard(color.colorString)
  }

  return (
    <ClickAwayListener onClickAway={handleTooltipClose}>
      <div>
        <Tooltip
          aria-label="foo-bar"
          disableFocusListener
          disableHoverListener
          disableTouchListener
          onClose={handleTooltipClose}
          open={isToolTipOpen}
          placement="left"
          PopperProps={{
            disablePortal: true,
          }}
          title="Copied!"
        >
          <IconButton
            aria-label={`Copy ${color.name} color`}
            onClick={handleButtonClick}
          >
            <FileCopyRounded />
          </IconButton>
        </Tooltip>
      </div>
    </ClickAwayListener>
  )
}
