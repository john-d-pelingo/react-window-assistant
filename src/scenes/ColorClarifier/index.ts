export { ColorClarifier as default } from './ColorClarifier'
