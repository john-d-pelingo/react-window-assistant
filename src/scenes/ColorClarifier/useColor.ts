import Color from 'color'
import { MutableRefObject, useEffect, useRef, useState } from 'react'

export const useColor = (colorText: string): MutableRefObject<Color | null> => {
  const color = useRef<Color | null>(null)
  const [, setIsColorValid] = useState<boolean>()

  useEffect(() => {
    try {
      color.current = new Color(colorText)
      setIsColorValid(true)
    } catch (error) {
      color.current = null
      setIsColorValid(false)
    }
  }, [colorText])

  return color
}
