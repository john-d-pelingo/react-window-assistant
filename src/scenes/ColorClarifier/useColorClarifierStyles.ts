import { createStyles, makeStyles } from '@material-ui/core'

export const useColorClarifierStyles = makeStyles(theme =>
  createStyles({
    container: {
      margin: `${theme.spacing(12.5)}px 0`,
    },
    invalidColorBox: {
      fontWeight: 'bold',
      marginTop: theme.spacing(3),
      textAlign: 'center',
    },
  }),
)
