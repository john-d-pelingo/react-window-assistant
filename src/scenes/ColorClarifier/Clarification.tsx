import {
  Paper,
  Table,
  TableBody,
  TableCell,
  TableHead,
  TableRow,
} from '@material-ui/core'
import Color from 'color'
import React, { FC } from 'react'
import styled from 'styled-components'

import { CopyColorButton } from './CopyColorButton'

const StyledPaper = styled(Paper)`
  width: 100%;
  margin-top: 24px;
  overflow-x: auto;
`

const ColorTableCell = styled(TableCell)`
  && {
    display: flex;
    align-items: center;
    justify-content: space-between;
  }
`

interface IClarificationProps {
  colorInstance: Color
}

export const Clarification: FC<IClarificationProps> = ({ colorInstance }) => {
  const colors: Array<{ colorString: string; name: string }> = [
    { colorString: colorInstance.rgb().toString(), name: 'RGB' },
    {
      colorString: colorInstance
        .hex()
        .toString()
        .toLowerCase(),
      name: 'Hex',
    },
    { colorString: colorInstance.hsl().toString(), name: 'HSL' },
    { colorString: colorInstance.hwb().toString(), name: 'HWB' },
  ]

  return (
    <StyledPaper>
      <Table
        style={{
          minWidth: 650,
        }}
      >
        <TableHead>
          <TableRow>
            <TableCell>Clarification</TableCell>
            <TableCell
              style={{
                backgroundColor: colorInstance.hex(),
              }}
            />
          </TableRow>
        </TableHead>
        <TableBody>
          {colors.map(({ colorString, name }, ii) => (
            <TableRow key={ii}>
              <TableCell>{name}</TableCell>
              <ColorTableCell aria-label={`${name} color`}>
                {colorString}
                <CopyColorButton
                  color={{
                    colorString,
                    name,
                  }}
                />
              </ColorTableCell>
            </TableRow>
          ))}
        </TableBody>
      </Table>
    </StyledPaper>
  )
}
