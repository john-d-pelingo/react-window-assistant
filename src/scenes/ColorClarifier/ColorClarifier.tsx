import { Container } from '@material-ui/core'
import { RouteComponentProps } from '@reach/router'
import React, { ChangeEvent, FC, useState } from 'react'

import { useColor } from 'hooks/userColor'

import { Clarification } from './Clarification'
import { useColorClarifierStyles } from './useColorClarifierStyles'

export const ColorClarifier: FC<RouteComponentProps> = () => {
  const classes = useColorClarifierStyles()
  const [colorText, setColorText] = useState('')

  const handleInputChange = (event: ChangeEvent<HTMLInputElement>): void => {
    setColorText(event.target.value)
  }

  const color = useColor(colorText.trim())

  return (
    <Container className={classes.container}>
      <input onChange={handleInputChange} type="text" value={colorText} />

      {/* TODO: append color query parameter on typing by reusing the function `appendQueryParameter` */}
      {color.current && <Clarification color={color.current} />}
    </Container>
  )
}
