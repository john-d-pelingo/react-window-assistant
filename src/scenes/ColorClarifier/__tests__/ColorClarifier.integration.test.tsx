import '@testing-library/jest-dom/extend-expect'
import 'jest-styled-components'

import { fireEvent, render } from '@testing-library/react'
import { createMemoryHistory } from 'history'
import React from 'react'
import { Helmet } from 'react-helmet'
import { Router } from 'react-router-dom'

import { ColorClarifier } from '../ColorClarifier'

let memoryHistory = createMemoryHistory()

describe('scenes - ColorClarifier', () => {
  beforeEach(() => {
    // NOTE: for the 'copies a color on color clarification click' test
    document.createRange = () => ({
      // @ts-ignore
      commonAncestorContainer: {
        nodeName: 'BODY',
        ownerDocument: document,
      },
      setEnd: () => undefined,
      setStart: () => undefined,
    })

    document.getSelection = jest.fn()
    document.execCommand = jest.fn()
    memoryHistory = createMemoryHistory()
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  it('mounts', () => {
    expect(memoryHistory.location.pathname).toBe('/')

    const { container } = render(
      <Router history={memoryHistory}>
        <ColorClarifier />
      </Router>,
    )

    expect(Helmet.peek().title).toBe('Color Clarifier')
    expect(container.firstChild).toMatchSnapshot()
  })

  it('displays info about a valid color', () => {
    expect(memoryHistory.location.pathname).toBe('/')

    const { getByLabelText, queryByText } = render(
      <Router history={memoryHistory}>
        <ColorClarifier />
      </Router>,
    )

    const colorInputElement = getByLabelText('Color input') as HTMLInputElement

    expect(document.activeElement).toEqual(colorInputElement)

    fireEvent.change(colorInputElement, {
      target: {
        value: 'hsl(291.29999999999995, 77%, 50.6%)',
      },
    })

    expect(colorInputElement.value).toBe('hsl(291.29999999999995, 77%, 50.6%)')
    expect(queryByText(/clarification/i)).toBeInTheDocument()
    expect(getByLabelText('RGB color').textContent).toBe('rgb(198, 32, 226)')
    expect(getByLabelText('Hex color').textContent).toBe('#c620e2')
    expect(getByLabelText('HSL color').textContent).toBe(
      'hsl(291.29999999999995, 77%, 50.6%)',
    )
    expect(getByLabelText('HWB color').textContent).toBe(
      'hwb(291.29999999999995, 12.6%, 11.4%)',
    )
    expect(memoryHistory.location.search).toBe(
      '?color=hsl%28291.29999999999995%2C%2077%25%2C%2050.6%25%29',
    )
  })

  it('displays that the color is invalid', () => {
    expect(memoryHistory.location.pathname).toBe('/')

    const { getByLabelText } = render(
      <Router history={memoryHistory}>
        <ColorClarifier />
      </Router>,
    )

    const colorInputElement = getByLabelText('Color input') as HTMLInputElement

    fireEvent.change(colorInputElement, {
      target: {
        value: 'Give me a color',
      },
    })

    expect(getByLabelText(/invalid input/i).textContent).toBe(
      'Invalid color value!',
    )
    expect(memoryHistory.location.search).toBe('?color=Give%20me%20a%20color')
  })

  it('copies a color on color clarification click', () => {
    expect(memoryHistory.location.pathname).toBe('/')

    const { getByLabelText, getByText } = render(
      <Router history={memoryHistory}>
        <ColorClarifier />
      </Router>,
    )

    const colorInputElement = getByLabelText('Color input') as HTMLInputElement

    fireEvent.change(colorInputElement, {
      target: {
        value: '#bca',
      },
    })

    const copyHwbButtonElement = getByLabelText(
      /copy hwb color/i,
    ) as HTMLButtonElement

    fireEvent.click(copyHwbButtonElement)

    const copiedTooltip = getByText(/copied!/i) as HTMLDivElement

    expect(copiedTooltip).toBeInTheDocument()
  })

  it('loads a correct color clarification from the URL', () => {
    expect(memoryHistory.location.pathname).toBe('/')

    memoryHistory.push('?color=%20rgb%2882%2C%20156%2C%2023%29')

    const { queryByText } = render(
      <Router history={memoryHistory}>
        <ColorClarifier />
      </Router>,
    )

    expect(queryByText(/clarification/i)).toBeInTheDocument()
  })

  it('loads a wrong color clarification from the URL', () => {
    expect(memoryHistory.location.pathname).toBe('/')

    memoryHistory.push('?color=kebap%20is%20ðŸ’¯')

    const { getByLabelText } = render(
      <Router history={memoryHistory}>
        <ColorClarifier />
      </Router>,
    )

    expect(getByLabelText(/invalid input/i)).toBeInTheDocument()
  })
})
