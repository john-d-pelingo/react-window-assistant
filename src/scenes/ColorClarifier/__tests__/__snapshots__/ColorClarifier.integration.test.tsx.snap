// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scenes - ColorClarifier mounts 1`] = `
<div
  class="MuiContainer-root MuiContainer-maxWidthXl"
  style="margin: 100px 0px;"
>
  <div
    class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
  >
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth Mui-focused Mui-focused MuiInputBase-formControl MuiInput-formControl"
    >
      <input
        aria-invalid="false"
        aria-label="Color input"
        class="MuiInputBase-input MuiInput-input"
        placeholder="Color to clarify"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;
